subsys/net/lib/dhcpv4/dhcpv4.c

```diff
@@ -810,6 +867,19 @@ static bool dhcpv4_parse_options(struct net_pkt *pkt,
 			break;
 		}
 #endif
+
+#if defined(CONFIG_NET_DHCPV4_OPTION_CALLBACKS_VENDOR_SPECIFIC)
+		case DHCPV4_OPTIONS_VENDOR_SPECIFIC: {
+
+			if (length >= 3) {
+				net_pkt_cursor_backup(pkt, &backup);
+				dhcpv4_parse_option_vendor(pkt, iface, msg_type, length);
+				net_pkt_cursor_restore(pkt, &backup);
```

Comment:
Why do we need to backup/restore cursor here? IMO `dhcpv4_parse_option_vendor()` should advance the cursor, as it parses consecutive option data. If it encounters problem (malformed option, for example no END marker), it should return an error.